### DISCUSSION TITLE
Difference between the `collision_constraint_config` and `collision_cost_config`

### DISCUSSION THREAD

**Discussion by mohamedsayed18** [tesseract-robotics/tesseract_planning#103] [Q&A] [answered]
I'm checking the [car seat example](https://github.com/ros-industrial-consortium/tesseract_ros/blob/master/tesseract_ros_examples/src/car_seat_example.cpp). My question is what is the difference between the collision cost and collision constraint?
for the cost:
```
trajopt_composite_profile->collision_cost_config.safety_margin = 0.005;
trajopt_composite_profile->collision_cost_config.coeff = 50;
```
As I understand, when the distance between the generated path and an object in the environment is less than the safety margin this path will be given a cost depend on the coeff. The goal is to get a path with the lowest cost.

How does it works for the constraint? If the distance between the generated path and an object in the environment is less than the safety margin this path is not acceptable?
In other words, the generated path must satisfy the condition(saftery_margin). Like in mathematical optimization we trying to optimize an objective function with respect to some variable in the presence of constraints

**Comment by Levi-Armstrong** [comment, marked as answer]
If you only set the collision as a cost it is not guaranteed to be collision-free the solver will just try to minimize it as best it can. If it is a constraint the solver will fail if any contact is within the safety margin. Now you may be wondering they why would you add it as a cost. In some complicated environments running the solver first with a collision as cost followed by running it again as a constraint performs better. Where if you added it as a constraint to start it may just fail because the initial state is in the infeasible region. I have also had good luck adding collision as both a cost and constraint with different safety margin distances.
